neuro_san/internals/tool_factories/service_tool_factory_provider.py

```diff
@@ -58,14 +73,32 @@ def add_agent_tool_registry(self, agent_name: str, registry: AgentToolFactory):
         """
         Register existing agent tool registry
         """
+        is_new: bool = False
         with self.lock:
-            is_new: bool = self.agents_table.get(agent_name, None) is None
+            is_new = self.agents_table.get(agent_name, None) is None
             self.agents_table[agent_name] = registry
+        # Notify listeners about this state change:
+        # do it outside of internal lock
+        for listener in self.listeners:
             if is_new:
-                self.logger.info("BUILT tool registry for agent %s", agent_name)
+                listener.agent_added(agent_name)
+                self.logger.info("ADDED tool registry for agent %s", agent_name)
             else:
+                listener.agent_modified(agent_name)
                 self.logger.info("REPLACED tool registry for agent %s", agent_name)
 
+    def remove_agent_tool_registry(self, agent_name: str):
+        """
+        Remove agent tool registry from the table
+        """
+        with self.lock:
+            self.agents_table.pop(agent_name, None)
+        # Notify listeners about this state change:
+        # do it outside of internal lock
+        for listener in self.listeners:
+            listener.agent_removed(agent_name)
+        self.logger.info("REMOVED tool registry for agent %s", agent_name)
+
```

Comment:
We may also remove agent tool registry from our state.
If we do, notify all registered listeners about this.
Note that we do that outside of internal lock protecting internal table - 
idea is to never run clients code inside our internal locks,
because you have no guarantees as to what their latency or side effects could be.